### PR TITLE
Add configurable zk node compression.

### PR DESCRIPTION
**Pull request by aquamatthias**
Compression information is stored as part of the surrounding protobuf.
Compression can be turned on/off via cmd line flags and is enabled by default.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/ZookeeperConf.scala [hunk 1]
```diff
@@ -35,6 +35,21 @@ trait ZookeeperConf extends ScallopConf {
     default = Some(25)
   )
 
+  lazy val zooKeeperCompressionEnabled = toggle("zk_compression",
+    descrYes = "Compression of zk nodes is enabled, if the size of the node is bigger than the configured threshold",
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/ZookeeperConf.scala]: Our flags description are usually written in imperative.

For example:

```
    descrYes = "(Default) Enable metric measurement of service method calls",
    descrNo = "Disable metric measurement of service method calls",
```

File: src/main/scala/mesosphere/marathon/ZookeeperConf.scala [hunk 2]
```diff
@@ -35,6 +35,21 @@ trait ZookeeperConf extends ScallopConf {
...
+    descrNo = "Compression of zk nodes is disabled",
+    noshort = true,
+    prefix = "disable_",
+    default = Some(true)
+  )
+
+  lazy val zooKeeperCompressionThreshold = opt[Long]("zk_compression_threshold",
+    descr = "Threshold in bytes, when compression is applied to the zk node.",
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/ZookeeperConf.scala]: I'ad add "(default: 64kb)" to the description.

File: src/main/scala/mesosphere/util/state/zk/ZKStore.scala [hunk 3]
```diff
@@ -115,17 +118,25 @@ case class ZKEntity(node: ZNode, data: ZKData, version: Option[Int] = None) exte
 }
 
 case class ZKData(name: String, uuid: UUID, bytes: IndexedSeq[Byte] = Vector.empty) {
-  def toProto: Protos.ZKStoreEntry = Protos.ZKStoreEntry.newBuilder()
-    .setName(name)
-    .setUuid(ByteString.copyFromUtf8(uuid.toString))
-    .setValue(ByteString.copyFrom(bytes.toArray))
-    .build()
+  def toProto(conf: CompressionConf): Protos.ZKStoreEntry = {
```
Comment (gkleiman) [inline src/main/scala/mesosphere/util/state/zk/ZKStore.scala]: I'd rename conf to compressionConf or even compression.

File: src/main/scala/mesosphere/util/state/zk/ZKStore.scala [hunk 4]
```diff
@@ -6,18 +6,21 @@ import com.fasterxml.uuid.impl.UUIDUtil
...
 import org.apache.log4j.Logger
 import org.apache.zookeeper.KeeperException
-import org.apache.zookeeper.KeeperException.{ NodeExistsException, NoNodeException }
+import org.apache.zookeeper.KeeperException.{ NoNodeException, NodeExistsException }
 
 import scala.concurrent.duration.Duration
 import scala.concurrent.{ Await, Future, Promise }
 
-class ZKStore(val client: ZkClient, rootNode: ZNode) extends PersistentStore {
+case class CompressionConf(enabled: Boolean, sizeLimit: Long)
```
Comment (gkleiman) [inline src/main/scala/mesosphere/util/state/zk/ZKStore.scala]: Our *Conf classes always inherit from ScallopCon. Is this one different for any special reason?

Maybe naming it CompressionSettings could make it clearer that it is not like the other *Conf classes we have.

File: src/test/scala/mesosphere/util/state/zk/ZKStoreTest.scala [hunk 5]
```diff
@@ -49,11 +49,37 @@ class ZKStoreTest extends PersistentStoreTest with StartedZookeeper with Matcher
     zkLoadUpdated.get.bytes should be("Hello again".getBytes)
   }
 
+  test("Entity nodes greater then the compression limit get compressed") {
```
Comment (gkleiman) [inline src/test/scala/mesosphere/util/state/zk/ZKStoreTest.scala]: s/then/than/

**Comment by gkleiman** [comment]
Looks pretty good to me.

Made just a few comments regarding the flag description and the CompressionConf class.

**Comment by gkleiman** [comment]
**Reminder:** we have to cherrypick this in master.

File: src/main/scala/mesosphere/util/state/zk/ZKStore.scala (hunk 4 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/util/state/zk/ZKStore.scala]: I don't want to use ScallopConf objects other than setup logic, mainly to make testing code easy.

**Comment by gkleiman** [comment]
We should add the new flags to our docs.

**Pull request merged**